src/charm.py

```diff
@@ -79,29 +117,38 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
                 return
 
         try:
+            logger.debug("opening tcp port")
             self._open_port_tcp(self._port)
         except subprocess.CalledProcessError:
             self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
             return
 
-        if not self._mongo.is_ready():
+        # check if this units deployment of MongoDB is ready
+        if not self._mongo.is_ready(all_replicas=False):
+            logger.debug("mongoDB not ready, defering on start event")
             self.unit.status = WaitingStatus("waiting for MongoDB to start")
             event.defer()
             return
 
-        # initialize replica set
-        self.unit.status = MaintenanceStatus("initialising MongoDB replicaset")
-        logger.debug("initialising replica set")
-        try:
-            logger.debug("initialising replica set for the following IPs %s", self._unit_ips)
-            self._mongo.initialize_replica_set(self._unit_ips)
-            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
-        except (ConnectionFailure, ConfigurationError) as e:
-            logger.error("error initialising replica sets in _on_start: error: %s", str(e))
-            self.unit.status = BlockedStatus("failed to initialise replicaset")
+        # mongod is now active
+        self.unit.status = ActiveStatus()
+
+        # only leader should initialise the replica set
+        if not self.unit.is_leader():
             return
 
-        self.unit.status = ActiveStatus()
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self.model.get_relation(PEER).units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s,"
+                + "waiting to initialise replica set",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
```

Comment:
If we `defer` here we will be in a situation where we forever wait for `self.app.planned_units() == len(self.model.get_relation(PEER).units) + 1` to be true. This is because the peer units won't be joined until the `relation-joined` hook is called. I have a more in depth explanation above for a similar situation in `relation-joined` let me tag you in it. :)

Reply:
Thanks for the explanation, Mia 😀. I understood this point. I'll just check your response for the last question from Daniela in the other comment.